### PR TITLE
Start foreground service on file open

### PR DESCRIPTION
**Pull request by wa2c**
https://github.com/wa2c/cifs-documents-provider/pull/53

Changed the foreground service starts when the file is opened.

@RebelliousX
Please check if the app works well

### CODE REVIEW

**Comment by RebelliousX** [comment]
@wa2c something wrong went in the last 2 commits.

I tested 8648cec1bf8e9322856c78efcbc274e6023aa8db and it works fine, although I get spammed with some notifications for the same opened file like 4 times And sometimes even after the emulator exits and I do force close it to be sure, the notifications are still there from CIFS Document Provider saying the files are still opened. SMBJ has the least issues compared to JCIFS-NG which I get lots of spammed notifications.

But f691a69e7fc322f96134e702498780cbb88cc255 (last commit), when I start playing lets say with PPSSPP, within 5 seconds, the emulator exits (not minimized, since if I open it again, I can't continue) and the notifications are still there.

**Comment by RebelliousX** [comment]
@wa2c I really appreciate your work and time for this.

Yesterday, I spent a great deal of time testing [b2a6d19](https://github.com/wa2c/cifs-documents-provider/pull/54/commits/b2a6d198dae6ed0fa101115158c15f81e73bdb80), specifically, fighting with the following code: 

```     
        fun cancel(service: Service, tag: String) {
        val notifications = notificationManager.activeNotifications.filter { it.id == NOTIFICATION_ID_PROVIDER }
        if (!notifications.any { it.tag == tag }) return

        notificationManager.cancel(tag, NOTIFICATION_ID_PROVIDER)
        if (notifications.size <= 2) {
            service.stopForeground(Service.STOP_FOREGROUND_REMOVE)
        }
    }
```

The problem is that "Open file" notification once it is canceled, the whole provider gets terminated, 

And today, I tried the changes in [48c10c6](https://github.com/wa2c/cifs-documents-provider/pull/54/commits/48c10c63725e4f540435a1bb2c8e688810a868f0). It is a shame we lost the ability to see which files are opened, but the situation did not improve much. 

"Open file" notification disappears while reading files' list randomly while I use PPSSPP, and it never comes back even when I try to run the game if I am lucky (the provider connections gets destroyed, I get hour glass in file picker indicating there is no connection, and I have to force close CIFS Doc Provider and try again, PPSSPP will forget the opened from CIFS Document Provider). But if the provider connections are still available, starting a game doesn't start the foreground service.

This might be unrelated, but I was reading somewhere in the documentation of Service class, that we can't start a foreground service in a background process or service starting with Android S (12)? I don't know if this has anything to do with this, but I thought of this because I think the provider is running in the background?

Again, I really appreciate your work and time to fix this. But I started to think that the solution I provided is the best since I don't have any issues with it, the service will never start if the option is not enabled, which I think is good. Please understand that it is not about "my solution", I really liked your idea, and I was hoping it will work, but I don't really think it will.

**Comment by RebelliousX** [comment]
@wa2c I think I found the root cause for this issue. 

I noticed something odd in the logcat. The problem is that file release is called as soon as the file is opened, So what happens is, when opening a file, the service is created, then it gets destroyed, then it tries to create it again within a span of 5 seconds or less. Android doesn't like that. For example, Android expects `startForeground()`  to be called (some articles say 5 seconds, others say within 20 seconds) right after `startForegroundService()` is called, otherwise, Android will stop the service and remove the notification.

Why file release callback is called when the file just opened, I don't know. This might be related to many issues in the issue tracker too, related to files not saving correctly?

Probably the implementation of `openDocument()` is kind of wrong here. Either we should use  `createReliablePipe()` or one of its related functions. Also, it might be worth looking at `ParcelFileDescriptor.OnCloseListener`, overriding `onClose()` when file descriptor is closed? It is actually the fourth parameter for `ParcelFileDescriptor.open(file, mode, handler, closelistener)`.

Anyways, in `CifsDocumentProvider.kt`. What I did to mitigate this, which seems to fix the issue, is by postponing the service stopping for some time. Lest another call for starting the service might happen just right afterwards, as follows:

```Java
    private fun stopFile(fileOpenId: String) {
        openingFileMap.remove(fileOpenId)
        // Postpone stopping service for 30 seconds..
        Timer().schedule( timerTask {
            if (openingFileMap.isEmpty()) {
                ProviderService.stop(providerContext)
            }
        }, 30000)
    }
```
I am still testing.. One thing I changed, which I am not sure if it helps, instead of `START_STICKY`, I used `START_NOT_STICKY`. According to this [article](https://medium.com/@engineermuse/foreground-services-in-android-e131a863a33d).

**Comment by RebelliousX** [comment]
Nice work @wa2c , I never thought about using workManager, I read slightly about it, this solution seems to be more elegant. It is running a foreground service under the hood.. Testing.. 👍😁

**Comment by RebelliousX** [comment]
@wa2c So far so good, I tested this thoroughly, it seems to be working well except for one small issue, which can be fixed easily.

I tested this and the workManager / foreground service is working fine. No issues here. I played over 1 hour of FFX and no disc read errors, using AetherSX2. And played some games on PPSSPP with my current favorite Tenchu Shadow Assassin😁 few missions in a single run, no issues.

The only problem I saw is that with AetherSX2, the notification doesn't get updated fast enough, PPSSPP doesn't have this issue. AetherSX2 from my debugging sessions, stores the list of all games and invoke the file by calling `queryDocument()` first, not `openDocument()`, weird. So when I play, the notification still says "Opening file" with no file mentioned in the notification. I tried changing the notification priority to high `IMPRTANCE_HIGH` but that did not help, so I changed it back to low.

But I checked `openingFileMap` container, and it does indeed contain the the ISO file of the game. Again, Android with its event system, it is not immediate in nature. What I did to fix this by calling `updateNotification()` again after 5-10 seconds. From my few tries, 5 seconds is a hit or miss, maybe because I am running a debug version, and it is not fast enough. But I changed it to 10 seconds and it works fine all the time.

`CifsDocumentProvider.kt`
```kotlin
private fun startFile(fileOpenId: String, uri: String) {
        logI("Start file: fileOpenid=$fileOpenId .. uri=$uri")
        if (openingFileMap.isEmpty()) {
            val request = OneTimeWorkRequest.Builder(ProviderWorker::class.java).build()
            workManager.enqueue(request)
        }
        openingFileMap[fileOpenId] = uri
        updateNotification()
        // force update the notifications again after 10 seconds.
        // just in case the first update call did not work correctly.
        Timer().schedule( timerTask {
            updateNotification()
        }, 10000)
    }
```
Please update this function, with this, you can disregard the `logI()`. I would consider the issue is fixed and ready to be merged 👍Unless you have something else to add of course 😁

File: data/src/main/java/com/wa2c/android/cifsdocumentsprovider/domain/repository/CifsRepository.kt [hunk 1]
```diff
@@ -50,16 +54,33 @@ class CifsRepository @Inject internal constructor(
...
+    private suspend fun removeBlockingQueue(dto: StorageConnection) {
         fileBlockingQueue.remove(dto)
+        updateOpeningFiles()
         logD("Queue removed: size=${fileBlockingQueue.count()}")
     }
+
+    private suspend fun updateOpeningFiles() {
+        _openUriList.emit(fileBlockingQueue.map { it.uri })
+    }
+
```
Comment (wa2c) [inline data/src/main/java/com/wa2c/android/cifsdocumentsprovider/domain/repository/CifsRepository.kt:83]: Manage open files

File: data/src/main/java/com/wa2c/android/cifsdocumentsprovider/domain/repository/CifsRepository.kt [hunk 2]
```diff
@@ -290,14 +311,15 @@ class CifsRepository @Inject internal constructor(
     /**
      * Get ProxyFileDescriptorCallback
      */
-    suspend fun getCallback(uri: String, mode: AccessMode): ProxyFileDescriptorCallback? {
+    suspend fun getCallback(uri: String, mode: AccessMode, onFileRelease: () -> Unit): ProxyFileDescriptorCallback? {
```
Comment (wa2c) [inline data/src/main/java/com/wa2c/android/cifsdocumentsprovider/domain/repository/CifsRepository.kt:314]: Use suspend function ( & fix storage client code)

File: presentation/src/main/java/com/wa2c/android/cifsdocumentsprovider/presentation/provider/CifsDocumentsProvider.kt [hunk 3]
```diff
@@ -65,7 +87,19 @@ class CifsDocumentsProvider : DocumentsProvider() {
...
+
     override fun onCreate(): Boolean {
+        logD("onCreate")
+        lifecycleOwner.start()
+
+        lifecycleOwner.lifecycleScope.launch {
+            cifsRepository.showNotification.collectIn(lifecycleOwner) {
+                updateWorker(it)
+            }
+        }
```
Comment (wa2c) [inline presentation/src/main/java/com/wa2c/android/cifsdocumentsprovider/presentation/provider/CifsDocumentsProvider.kt:101]: Show/hide notification

File: presentation/src/main/java/com/wa2c/android/cifsdocumentsprovider/presentation/provider/ProviderWorker.kt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        logD("ProviderWorker begin")
+
+        try {
+            lifecycleOwner.start()
+            lifecycleOwner.lifecycle.coroutineScope.launch {
+                cifsRepository.openUriList.collectIn(lifecycleOwner) { list ->
+                    providerNotification.updateFiles(list)
+                }
+            }
+            setForeground(providerNotification.getNotificationInfo(cifsRepository.openUriList.value))
```
Comment (wa2c) [inline presentation/src/main/java/com/wa2c/android/cifsdocumentsprovider/presentation/provider/ProviderWorker.kt:39]: Update notification file list.

File: data/src/main/java/com/wa2c/android/cifsdocumentsprovider/domain/repository/CifsRepository.kt [hunk 5]
```diff
@@ -50,16 +54,33 @@ class CifsRepository @Inject internal constructor(
...
     }
 
+    /** Connected file uri list */
+    private val _openUriList = MutableStateFlow<List<String>>(emptyList())
+    val openUriList = _openUriList.asStateFlow()
+
+    /** Show notification  */
+    val showNotification: Flow<Boolean> = openUriList.map {
+        it.isNotEmpty() && appPreferences.useForegroundFlow.first()
+    }.distinctUntilChanged()
```
Comment (wa2c) [inline data/src/main/java/com/wa2c/android/cifsdocumentsprovider/domain/repository/CifsRepository.kt:64]: Show/Hide notification flag flow.

**Review by wa2c — COMMENTED** [review]
@RebelliousX 
I fixed the notification updater. ( by flow and observer. )

I renamed "ForegroundService" to "Foreground" in all sources. 
So notification visible settings key changed.

I think the current source is working fine. Please use all means.
If it has no problem, I merge to main branch.

**Review by RebelliousX — APPROVED** [review]
PERFECT!! Thank you so so much.  I tested the latest changes and indeed the notifications are fixed. The process doesn't die like before randomly, and even it is a cleaner approach than using "foreground service" since sometimes when we close the foreground service it closes / destroys the provider, that happened rarely. But now everything is good.

Besides this PR fixes Issue #52. On a positive side effects, somewhere along these past 15 commits in this PR also fixed issue #49 too. The `addBlockingQueue()` doesn't seem to get full and I was able to fully scan games' library using AetherSX2. Before it was getting stuck at file limite - 1 or 2. I showed you a video before about it if you remember.

Finally, unrelated to this, please fix the release apk build! This started happening from 8f4017318604e992d6d552d6247081ec0ff64e8b, I reverted the changes of the files from that commit and the release apk builds fine.

File: data/src/main/java/com/wa2c/android/cifsdocumentsprovider/domain/repository/CifsRepository.kt (hunk 2 above)
Comment (RebelliousX) [reply data/src/main/java/com/wa2c/android/cifsdocumentsprovider/domain/repository/CifsRepository.kt:314]: Yup, I believe this fixed file release issue of being released when it was just opened.

File: presentation/src/main/java/com/wa2c/android/cifsdocumentsprovider/presentation/provider/CifsDocumentsProvider.kt (hunk 3 above)
Comment (RebelliousX) [reply presentation/src/main/java/com/wa2c/android/cifsdocumentsprovider/presentation/provider/CifsDocumentsProvider.kt:101]: Using observer to monitor changes of opened URI list is an excellent approach.  I actually, yesterday, asked chat GPT how to monitor changes to a variable in kotlin and call a function when that happened. The answer was by using Delegates. But I didn't give it much of a thought afterwards.

File: presentation/src/main/java/com/wa2c/android/cifsdocumentsprovider/presentation/provider/ProviderWorker.kt (hunk 4 above)
Comment (RebelliousX) [reply presentation/src/main/java/com/wa2c/android/cifsdocumentsprovider/presentation/provider/ProviderWorker.kt:39]: This was what causing the notification not displaying anything when opening a game in AetherSX2. The list was empty and create notification function creates an empty notification with nothing but "opening file". updating that was not fast enough. 

But now with this, the updated list is done inside `doWork()` for workManager. Nice 👍

**Review by RebelliousX — APPROVED** [review]
Nicely done. Everything is good.

**Comment by wa2c** [comment]
@RebelliousX
Thank you for your help! 
I merge into main branch

**Pull request merged**